### PR TITLE
Geojson simplify

### PR DESCRIPTION
**Pull request by fpernice-google**
Script to simplify the data downloaded by the script in PR #192 .

### CODE REVIEW

File: tools/geojson_simplifier/simplify_test.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import unittest
+import geojson
+import simplify
+
+TEST_DATA_DIR = "test-data/"
+
+
+class GeojsonSimplifierTest(unittest.TestCase):
+    def test_california(self):
```
Comment (jeffreyoldham) [inline tools/geojson_simplifier/simplify_test.py:29]: \# TODO(fpernice-google): Add some simple tests using hard-coded values including the empty input, a single input, and a few inputs.

File: tools/geojson_simplifier/simplify_test.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        simplifier = simplify.GeojsonSimplifier()
+        simplifier.read_geojson(TEST_DATA_DIR + 'cali.geojson')
```
Comment (jeffreyoldham) [inline tools/geojson_simplifier/simplify_test.py:31]: os.path.join(TEST_DATA_DIR, 'cali.geojson')

This will work on more operating systems, eliminating the assumption that "/" is the file separator.  Also, revise TEST_DATA_DIR to not end with a slash.  Thanks.

File: tools/geojson_simplifier/simplify.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        original_sz = sum([len(coords[i][0]) for i in range(len(coords))])
+        if verbose:
+            print(f"Original number of points = {original_sz}.")
+
+        for i in range(len(coords)):
+            assert len(coords[i]) == 1
+            c = coords[i][0]
+            new_c = rdp(c, epsilon=0.01)
+            if len(new_c) >= 3:
```
Comment (jeffreyoldham) [inline tools/geojson_simplifier/simplify.py:63]: \# Simplify the polygon succeeded, not yielding a line

File: tools/geojson_simplifier/simplify.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    def simplify(self, epsilon=0.01, verbose=False):
+        self.simple_geojson = self.raw_geojson.copy()
+        coords = self.simple_geojson['coordinates']
+
+        original_sz = sum([len(coords[i][0]) for i in range(len(coords))])
+        if verbose:
+            print(f"Original number of points = {original_sz}.")
+
+        for i in range(len(coords)):
```
Comment (jeffreyoldham) [inline tools/geojson_simplifier/simplify.py:59]: Iterate over multipolygons.

File: tools/geojson_simplifier/simplify.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            print(f"Simplified number of points = {simplified_sz}.")
+
+    def save(self, out_path):
+        with open(out_path, 'w') as f:
+            geojson.dump(self.simple_geojson, f)
+
+
+def main(_):
+    flags.mark_flag_as_required('in_path')
+    flags.mark_flag_as_required('out_path')
```
[the remark this answers is not in the document]
Comment (fpernice-google) [reply tools/geojson_simplifier/simplify.py:77]: Moved. I had looked at an example that put them here but it seems like in the abseil documentation they are placed on top.

File: tools/geojson_simplifier/simplify.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+        Args:
+            epsilon: The epsilon parameter to the  Ramer–Douglas–Peucker
+                     algorithm. See the Wikipedia page for details.
+            verbose: If True, the number of points in the GeoJSON before and
+                     after simplification will be printed for comparison.
+        """
+        coords = self.geojson['coordinates']
+
+        original_sz = 0
```
Comment (beets) [inline tools/geojson_simplifier/simplify.py:86]: What does sz mean?

File: tools/geojson_simplifier/simplify.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        """
+        with open(in_path, 'r') as f:
+            self.geojson = geojson.load(f)
+            if not self.geojson.is_valid:
+                raise ValueError("Invalid GeoJSON read in.")
+
+    def simplify(self, epsilon=0.01, verbose=False):
+        """Modifies the instance geojson by reducing its number of points.
+
+        Runs the Ramer–Douglas–Peucker algorithm to simplify the GeoJSONs.
```
Comment (beets) [inline tools/geojson_simplifier/simplify.py:76]: Can you link to your reference for the algorithm?

File: tools/geojson_simplifier/simplify.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            original_sz += len(c)
+            new_c = rdp.rdp(c, epsilon=epsilon)
+            simplified_sz += len(new_c)
+            if len(new_c) >= 3:
+                # Simplify the polygon succeeded, not yielding a line
+                coords[i][0] = new_c
+
+        if verbose:
+            print(f"Original number of points = {original_sz}.")
+        if verbose:
```
Comment (beets) [inline tools/geojson_simplifier/simplify.py:101]: Combine this with the above if-clause?

File: tools/geojson_simplifier/simplify_test.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        simplifier.read_geojson(os.path.join(TEST_DATA_DIR,
+                                             'california.geojson'))
+        simplifier.simplify()
+        result = simplifier.geojson
+        file = os.path.join(TEST_DATA_DIR, 'california-simple.geojson')
+        with open(file, 'r') as f:
+            expected_result = geojson.load(f)
+        self.assertDictEqual(result, expected_result)
+
+    def test_alabama(self):
```
Comment (beets) [inline tools/geojson_simplifier/simplify_test.py:40]: These two tests contain identical code, just different comparison data. Could you refactor some code out? What's important to do after that is to ensure that the message that gets printed out on failure tells you which file caused the issue (otherwise it's hard to know which case failed).

**Comment by beets** [comment]
README in a separate PR SGTM.

**Comment by beets** [comment]
@shifucun the stored maps here are used only for the test :)

**Comment by fpernice-google** [comment]
> I don't have much context on this PR, can you detailed how this will be used in the Choropleth tool? In general I am puzzled on why you want to store the actual data here.

Answered in https://github.com/datacommonsorg/website/pull/192#issuecomment-666738564
